tc/core/polyhedral/cuda/mapped_scop.cc

```diff
@@ -531,24 +543,25 @@ isl::multi_aff constructThreadToWarp(
 }
 } // namespace
 
-isl::multi_union_pw_aff MappedScop::threadMappingSchedule(
+template <typename MappingTypeId>
+isl::multi_union_pw_aff MappedScop::mappingSchedule(
```

Comment:
This should be a free function with static scop that takes the parameters it needs.
Exposing its declaration is useless (since it's not supposed to be used), error-prone (if used with some non-explicitly instantiated type in the future -> linking errors) and increases the API.

Alternatively it could go to a -inl.h and we could drop both `threadMappingSchedule` and `blockMappingSchedule`; whichever is fine with me.

Reply:
On Tue, Jul 03, 2018 at 11:43:23AM -0700, Nicolas Vasilache wrote:
> This should be a free function with static scop that takes the parameters it needs.

What exactly do you propose?
(I don't see why `MappedScop::mappingSchedule` should have a different
prototype from `MappedScop::threadMappingSchedule`.)

> Exposing its declaration is useless (since it's not supposed to be used), error-prone (if u
sed with some non-explicitly instantiated type in the future -> linking errors) and increases
 the API.

The declaration is not useless.  It is used in the last commit.
As explained in the commit message, there are only two possible instantiations.

> Alternatively it could go to a -inl.h and we could drop both `threadMappingSchedule` and `b
lockMappingSchedule`; whichever is fine with me.

I don't see the point.  I'll just move it into `tighten_launch_bounds.cc`.

skimo

Reply:
> if used with some non-explicitly instantiated type in the future -> linking errors

`static_assert` to the rescue!

I think @nicolasvasilache looks for something like `collectMappingsTo` that we have in memory promotion heuristic code https://github.com/facebookresearch/TensorComprehensions/blob/7f4ed011b9df9695cf7b8b366d66d90ab2685185/tc/core/polyhedral/cuda/memory_promotion_heuristic.cc#L303

`threadMappingSchedule` comes from my commit, I believe.  The idea was to have `mappingSchedule<mapping::Block>()` and `mappingFilter<mapping::Block>()` together at some point, they only differ by the return type.